merkle/hstar2_test.go

```diff
@@ -28,59 +29,56 @@ import (
 
 const treeID = int64(0)
 
-// This root was calculated with the C++/Python sparse Merkle tree code in the
+// Some known answers for incrementally adding index/value pairs to a sparse tree.
+// rootB64 is the incremental root after adding the corresponding i/v pair, and
+// all i/v pairs which come before it.
+//
+// Values were calculated with the C++/Python sparse Merkle tree code in the
 // github.com/google/certificate-transparency repo.
 // TODO(alcutter): replace with hash-dependent computation. How is this computed?
-var sparseEmptyRootHashB64 = testonly.MustDecodeBase64("xmifEIEqCYCXbZUz2Dh1KCFmFZVn7DUVVxbBQTr1PWo=")
-
-// createHStar2Leaves builds a list of HStar2LeafHash structs suitable for
-// passing into a the HStar2 sparse Merkle tree implementation.
-// The map keys will be SHA256 hashed before being added to the returned
-// structs.
-func createHStar2Leaves(treeID int64, hasher hashers.MapHasher, values map[string]string) []HStar2LeafHash {
-	r := []HStar2LeafHash{}
-	for k := range values {
-		khash := testonly.HashKey(k)
-		vhash := hasher.HashLeaf(treeID, khash, hasher.BitLen(), []byte(values[k]))
-		r = append(r, HStar2LeafHash{
-			Index:    new(big.Int).SetBytes(khash),
-			LeafHash: vhash[:],
-		})
-	}
-	return r
+var simpleTestVector = []struct {
+	index, value, root []byte
+}{
+	{nil, nil, b64("xmifEIEqCYCXbZUz2Dh1KCFmFZVn7DUVVxbBQTr1PWo=")}, // Empty tree.
+	{testonly.HashKey("a"), []byte("0"), b64("nP1psZp1bu3jrY5Yv89rI+w5ywe9lLqI2qZi5ibTSF0=")},
+	{testonly.HashKey("b"), []byte("1"), b64("EJ1Rw6DQT9bDn2Zbn7u+9/j799PSdqT9gfBymS9MBZY=")},
+	{testonly.HashKey("a"), []byte("2"), b64("2rAZz4HJAMJqJ5c8ClS4wEzTP71GTdjMZMe1rKWPA5o=")},
 }
 
-func TestHStar2EmptyRootKAT(t *testing.T) {
-	s := NewHStar2(treeID, maphasher.Default)
-	root, err := s.HStar2Root(s.hasher.Size()*8, []HStar2LeafHash{})
-	if err != nil {
-		t.Fatalf("Failed to calculate root: %v", err)
+// createHStar2Leaves returns a []HStar2LeafHash formed by the mapping of index, value ...
+// createHStar2Leaves panics if len(iv) is odd. Duplicate i/v pairs get over written.
```

Comment:
This algorithms is inspired by gRPC's `[metadata.Pairs](https://github.com/grpc/grpc-go/blob/master/metadata/metadata.go#L65)` which also has a key / value pair format. 

I use a map internally to also guarantee that emitted values don't have duplicate keys.